### ISSUE TITLE
julia_setup() doesn't work

### ISSUE THREAD

**Issue by JianghuiDu** [Non-Contradiction/JuliaCall#72]
Windows 10, R 3.5.1, Julia 1.0
I installed the development version of `JuliaCall` and when it reports error. 
```
> library(JuliaCall)
> julia <- julia_setup()
Julia version 1.0.0 at location C:\Users\JD\AppData\Local\Julia-1.0.0\bin will be used.
Error in juliacall_initialize(.julia$dll_file) : 
  jl_init - The specified procedure could not be found.
```
The installation process was sucessful. `Julia` is on my windowns `PATH` and I checked it from the command line. Any help?

**Comment by Non-Contradiction** [comment]
The error seems weird. I know where it comes from, but it shouldn't have happened. And it seems like an error from previous versions of JuliaCall.

How about
* download the source of JuliaCall from Github,
* open JuliaCall.Rproj in your RStudio or open R from the directory where you download the source of JuliaCall,
* in the `src/libjulia.cpp`, find the line:
  https://github.com/Non-Contradiction/JuliaCall/blob/0a66642c807f0be64e0638191a7427cff156302e/src/libjulia.cpp#L120
  and change it to `if (!(load_symbol(libjulia_t, "jl_init__threading", (void**) &jl_init)))`
* Rebuild the package, you can use the build panel on the upper right corner of RStudio and click the `Install and Restart` button.

And then see if the error persists.

**Comment by JianghuiDu** [comment]
I was able to rebuilt it sucessfully, but once I call `julia_setup()` the session is aborted. I tried a few times and it all get aborted.

**Comment by Non-Contradiction** [comment]
Does it have any error message before the session is aborted?

**Comment by JianghuiDu** [comment]
No. It didn't report anything, just terminated the session outright in Rstudio. I also tried in R's native console, in which case it's stuck (again no error) and nothing happened for half an hour and I have to close it.

**Comment by Non-Contradiction** [comment]
How about trying it from the command line, or use the `Check` button besides the `Install and Restart` button in RStudio, which should be equivalent to running from command line.

**Comment by JianghuiDu** [comment]
Same here. All the checks are passed. I just followed @metanoid  and found that I also can run `julia_setup()` from windows command line, but not from R console or Rstudio.

**Comment by JianghuiDu** [comment]
Here are the `check` results if it helps
```
==> devtools::check(document = FALSE)

Setting env vars --------------------------------------------------------------
CFLAGS  : -Wall -pedantic
CXXFLAGS: -Wall -pedantic
Building JuliaCall ------------------------------------------------------------
"C:/PROGRA~1/R/R-3.5.1/bin/x64/R" --no-site-file --no-environ --no-save  \
  --no-restore --quiet CMD build "C:\Users\JD\Downloads\JuliaCall-master"  \
  --no-resave-data --no-manual 

* checking for file 'C:\Users\JD\Downloads\JuliaCall-master/DESCRIPTION' ... OK
* preparing 'JuliaCall':
* checking DESCRIPTION meta-information ... OK
* cleaning src
* installing the package to build vignettes
* creating vignettes ... OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building 'JuliaCall_0.16.0.9000.tar.gz'

Setting env vars --------------------------------------------------------------
_R_CHECK_CRAN_INCOMING_USE_ASPELL_: TRUE
_R_CHECK_CRAN_INCOMING_           : FALSE
_R_CHECK_FORCE_SUGGESTS_          : FALSE
Checking JuliaCall ------------------------------------------------------------
"C:/PROGRA~1/R/R-3.5.1/bin/x64/R" --no-site-file --no-environ --no-save  \
  --no-restore --quiet CMD check  \
  "C:\Users\JD\AppData\Local\Temp\Rtmpa2zB1A/JuliaCall_0.16.0.9000.tar.gz"  \
  --as-cran --timings --no-manual 

* using log directory 'C:/Users/JD/Downloads/JuliaCall.Rcheck'
* using R version 3.5.1 (2018-07-02)
* using platform: x86_64-w64-mingw32 (64-bit)
* using session charset: ISO8859-1
* using options '--no-manual --as-cran'
* checking for file 'JuliaCall/DESCRIPTION' ... OK
* checking extension type ... Package
* this is package 'JuliaCall' version '0.16.0.9000'
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking serialization versions ... OK
* checking whether package 'JuliaCall' can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking 'build' directory ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking loading without being on the library search path ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking line endings in C/C++/Fortran sources/headers ... OK
* checking pragmas in C/C++ headers and code ... OK
* checking compilation flags used ... OK
* checking compiled code ... OK
* checking installed files from 'inst/doc' ... OK
* checking files in 'vignettes' ... OK
* checking examples ... OK
** found \donttest examples: check also with --run-donttest
* checking for unstated dependencies in 'tests' ... OK
* checking tests ...
  Running 'testthat.R' OK
 [52s]
* checking for unstated dependencies in vignettes ... OK
* checking package vignettes in 'inst/doc' ... OK
* checking re-building of vignette outputs ... OK
* DONE
Status: OK



R CMD check results
0 errors | 0 warnings | 0 notes

R CMD check succeeded
```
Seems like a strange problem.

**Comment by Non-Contradiction** [comment]
Thank you for the feedback!
R 32-bit can only work with Julia 32-bit and R 64-bit can only work with Julia 64-bit. 
So it is normal that the 32-bit tests failed, since the Julia is 64-bit.

Since the problem seems to be with RStudio and R GUI and the appveyor CI testing cannot test GUI things, I will try to find Windows machines for testing.

**Comment by JianghuiDu** [comment]
Mine (R and Julia) are both 64-bit. Rstudio is 32-bit though. Hope there will be a solution. Thanks.

**Comment by Non-Contradiction** [comment]
I have a somewhat crazy idea. How about running RStudio from the command line and then do the `library(JuliaCall); julia_setup()` in the RStudio?
I came up with this idea because I found an issue here <https://github.com/JuliaLang/julia/issues/28820> that on windows, embedded julia 1.0 will crash if there is no console (command line).
Since the problem here is common to both R GUI and RStudio, but not seen from the command line, it seems that these problems may be related.

**Comment by JianghuiDu** [comment]
If you mean to start RStudio from the command line rather than the launch button then it doesn't make a difference. Did I misunderstand you?

**Comment by Non-Contradiction** [comment]
No misunderstanding, this is exactly what I mean.
New version of RStudio has a terminal. How about running R from the RStudio terminal and then trying in this R?

**Comment by JianghuiDu** [comment]
Yes, it works as it does using windows terminal. I mean the RStudio terminal is simply calling the windows terminal so they should behave the same, right?

**Comment by Non-Contradiction** [comment]
Yes, it should be the same. And I found that RStudio starts a new session for each terminal, so this cannot solve the real issue, but it does provide a temporary solution: starting an R in the RStudio terminal and then use RStudio's `Code` ->  `Send to Terminal` to run R codes in the terminal R instead of RStudio's console.

The issue I linked to in my previous post does have a solution (?I'm not sure), but it requires modifying julia source code and build julia yourself, which seems too tedious  to try...

**Comment by JianghuiDu** [comment]
That would work but it would be much simpler just doing julia in julia and then export results to be read into R. Thanks. If some simple solution comes up I would appreciate to learn. Thanks.

**Comment by cyclesix** [comment]
Just to report, I experience this same issue on 2 different computers using both R 3.5.1 and R 3.4.4:
As soon as I call julia_setup() in RStudio or R GUI session crashes.

I am able to run julia_setup() as expected in R from terminal (cmd).

**Comment by Non-Contradiction** [comment]
The issue https://github.com/JuliaLang/julia/issues/28820 has been fixed on Julia master and should now be included in the Julia windows nightly binary at <https://julialang.org/downloads/nightlies.html>, the nightly is not stable (but the mac nightly works on my mac without obvious problems) and maybe we can use it to check whether the fix solves the issue here. 

If the RStudio and R GUI issue is solved with the nightly, then https://github.com/JuliaLang/julia/issues/28820 is the root of the problem here, and we can expect the fix will be in the Julia patch release like 1.0.2 next month (hopefully). If not, then the root of the problem here is somewhere else, and I need to investigate more into the problem.

**Comment by JianghuiDu** [comment]
After installing Julia windows nightly both my Rstudio and Julia 1.0 stopped working immediately. They just froze and restarting/reinstalling didn't help. I have to uninstall the nightly and everything came back normal...

**Comment by Non-Contradiction** [comment]
@JianghuiDu The nightly Julia itself doesn't work either? Is there anything related to Julia in PATH variable of windows? Maybe we should try to rebuild JuliaCall?

**Comment by JianghuiDu** [comment]
The nightly itself actually worked. Everything else stopped. I wasn't even doing anything with JuliaCall in R yet still my Rstudio stopped working. It's really strange.

**Comment by samclifford** [comment]
I find myself in the same situation. Loads fine on my MacBook but on my Windows machine, RStudio announces the session has been aborted after the Console shows the message that `Julia version 0.7.0 at location c:\JULIA-~1.0\bin will be used`. R 3.5.1 on Windows 10, with RStudio 1.1.456. Same thing happens with Julia 1.0.0.

Upgrading to the nightly Julia 1.1.0-DEV.535 (Commit 84024a1f44) seems to have resolved the issue.

**Comment by Non-Contradiction** [comment]
@samclifford Thanks for the feedback! Then the problem root is indeed the Julia issue when embedding on windows, Julia nightly ( and possibly Julia 1.0.2 later?) will fix the RStudio issue.

**Comment by samclifford** [comment]
Julia nightly stopped working after a while. Not sure what's happened there. Testing on a Linux box now.

**Comment by statli** [comment]
![1](https://user-images.githubusercontent.com/34979978/47896581-20724880-dea9-11e8-8fc6-ff6acfe47df3.png)

**Comment by Non-Contradiction** [comment]
@statli Thanks for reporting! Does it work in R starting from command line? As discussed before, this is likely due to an issue in Julia itself, which crashes the GUI application on windows when embedding, and will (hopefully) be fixed in Julia 1.0.2 as far as I know.

**Comment by statli** [comment]
@Non-Contradiction .Thanks for your answer! This is my Windows OS .It does not work in R starting from command line,have same problem.Now,I ues JuliaCall package in CentOS. The package is very good. Thank you again for your work.

**Comment by statli** [comment]
@Non-Contradiction Now,my problem become[https://github.com/Non-Contradiction/JuliaCall/issues/67](url)
![21](https://user-images.githubusercontent.com/34979978/47948369-f8e3b480-df6a-11e8-9873-82938b5013f0.png)

**Comment by Non-Contradiction** [comment]
@statli The new problem can be fixed by setting the PATH variable in Windows to contain the path for julia binary, or `setwd` to the path of julia binary when doing `julia_setup()`, but I'm afraid once this problem is fixed, the old problem will still occur that julia crashes RStudio.

What is the error message for the R starting from command line? As far as I know, starting R from command line should be fine, or even from RStudio's terminal (not RStudio's console), or from the Rmd document using the "knit" button, because all these three methods are actually the same in invoking a command line (terminal) R.

**Comment by statli** [comment]
@Non-Contradiction Thanks,sir.

**Comment by Non-Contradiction** [comment]
Dear all,
Julia 1.0.2 just comes out! As said before, the RStudio crash is likely to be caused by a Julia issue, which should be fixed in Julia 1.0.2.
Thanks again for providing the feedbacks!

**Comment by jmerc13** [comment]
Hi,
I was able to setup JuliaCall using 1.0.2 only after adding a new PATH variable on my Windows 10 64-bit machine. Just FYI.

**Comment by JianghuiDu** [comment]
After updating to 1.0.2 everything works fine (yes, need to set up path first).

**Comment by ChrisRackauckas** [comment]
I haven't been able to get v1.0.2 to work even with setting the path. Maybe I'm doing something wrong?

```R
> library(diffeqr)
> diffeqr::diffeq_setup()
Error in JuliaCall::julia_setup(...) : Julia is not found.
> library(JuliaCall)
> julia_setup(JULIA_HOME = "C:\Users\Chris\AppData\Local\Julia-1.0.2\bin")
Error: '\U' used without hex digits in character string starting ""C:\U"
> julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin")
Error in julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin") : 
  Julia is not found.
> julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin\\julia")
Error in julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin\\julia") : 
  Julia is not found.
```

Is it looking for `Julia` not `julia`?

**Comment by Non-Contradiction** [comment]
@ChrisRackauckas `Julia` is just the printing message, which should have no effect. `JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin"` should be fine. 
Maybe you could try to 
```
originwd <- getwd(); 
setwd(C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin)
julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin")
setwd(originwd)
```
The master of JuliaCall does the `setwd` thing automatically on windows, but the released version doesn't.

**Comment by ChrisRackauckas** [comment]
```
> setwd(originwd)
> originwd <- getwd(); 
> setwd("C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin")
> julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin")
Error in julia_setup(JULIA_HOME = "C:\\Users\\Chris\\AppData\\Local\\Julia-1.0.2\\bin") : 
  Julia is not found.
```

**Comment by Non-Contradiction** [comment]
@ChrisRackauckas I have never seen this problem before. Is `JULIA_HOME` also in Windows PATH variable? I think if `setwd` is used, then there is no need for the PATH variable, but just want to make sure. And this problem seems to be worth a new issue.

**Comment by MarieEtienne** [comment]
Hello,
I have the same kind of message with Ubunutu 18.04, Rstudio 1.1.456 and R 3.5.1. I have installed Julia with snap, but I got the following error with julia_setup() commandJulia version 1.0.4 at location /snap/julia/13/bin will be used.
Error in dyn.load(.julia$dll_file) : 
  unable to load shared object '/snap/julia/13/bin/../lib/libjulia.so.1':
  /snap/julia/13/bin/../lib/libjulia.so.1: wrong ELF class: ELFCLASS32
Error in juliacall_initialize(.julia$dll_file) : 
  /snap/julia/13/bin/../lib/libjulia.so.1 - /snap/julia/13/bin/../lib/libjulia.so.1: wrong ELF class: ELFCLASS32
I tried to change the line you mentionned in libjulia.cpp, but the error is still the same.

Thanks for your help

**Comment by Non-Contradiction** [comment]
Hi @MarieEtienne, yours seems to be a separate issue. Sine the

**Comment by Non-Contradiction** [comment]
Hi @MarieEtienne, yours seems to be another issue. Before any further investigation, you should not change the code in libjulia.cpp. 
From the error message, it seems that the problem is related to 32bit v.s. 64bit. How about checking `versioninfo()` directly from Julia and see `sessioninfo()` in R?

**Comment by MarieEtienne** [comment]
You got it correctly. snap has installed a 32bits version. Changing to a 64 bits version for Julia (so is my installed  R version) solve it. Thanks a lot  @Non-Contradiction  for your  super quick answer.

**Comment by stanleesocca** [comment]
I have some issue setting up the julia in my system: 
````
library(JuliaCall)
> julia <- JuliaCall::julia_setup(JULIA_HOME = "C:\\Users\\nmors\\AppData\\Local\\Programs\\Julia 1.5.0\\bin")
Julia version 1.5.0 at location C:\Users\nmors\AppData\Local\Programs\JULIA1~1.0\bin will be used.
Loading setup script for JuliaCall...
Error in .julia$cmd(paste0("Base.include(Main,\"", system.file("julia/setup.jl",  : 
  Error happens when you try to execute command Base.include(Main,"C:/Users/nmors/Documents/R/win-library/4.0/JuliaCall/julia/setup.jl") in Julia.
````

Any idea how to solve it.

**Comment by Non-Contradiction** [comment]
@stanleesocca 
From the current information, it is hard to diagnose the problem. You can try to
* Find R_HOME by `R.home()` in R.
* Run `ENV["R_HOME"] = "the path to the R_HOME"` in Julia.
* Run `Base.include(Main,"C:/Users/nmors/Documents/R/win-library/4.0/JuliaCall/julia/setup.jl")` in Julia as suggested in the error message to have more detailed error information.

I have some guess. Maybe the problem is due to some upgrade in R? If that is the case, then `julia_setup("C:\\Users\\nmors\\AppData\\Local\\Programs\\Julia 1.5.0\\bin", rebuild = TRUE)` could help.

**Comment by chancejohnstone** [comment]
Just an FYI, I updated R to version 4.1.0 and had issues with julia_setup(). It continued to hang at "Loading setup script for JuliaCall."  I tried to install and setup multiple versions of julia (1.0.5, 1.4.2) and nothing worked. Changed back to R-4.0.1 and there were no issues. Don't know if this is helpful, but I thought I would pass on the info.

**Comment by Non-Contradiction** [comment]
@chancejohnstone 
With a new R version, maybe using `rebuild = TRUE` in `julia_setup()` can help.
